File: packages/validation/quality.py
"""Heuristic quality scoring for flashcards.

Implements a 5-dimension rubric: clarity, atomicity, testability,
memorability, accuracy.  Each dimension scores 0.0-1.0.
"""

from __future__ import annotations

import re
from dataclasses import dataclass

_VAGUE_PATTERNS: tuple[re.Pattern[str], ...] = tuple(
    re.compile(p, re.IGNORECASE)
    for p in (
        r"^explain\s+",
        r"^describe\s+",
        r"^tell\s+(me\s+)?about\s+",
        r"^discuss\s+",
        r"^elaborate\s+(on\s+)?",
    )
)

_YES_NO_STARTERS: tuple[str, ...] = (
    "is ", "are ", "does ", "do ", "can ", "will ", "has ", "have ",
    "was ", "were ", "did ", "could ", "would ", "should ",
)

_MULTI_CONCEPT_RE = re.compile(r"\b(and|or)\b", re.IGNORECASE)
_ENUM_RE = re.compile(r"^\s*(?:[-*]|\d+[.)])\s", re.MULTILINE)


@dataclass(frozen=True, slots=True)
class QualityScore:
    """Five-dimension quality assessment of a flashcard."""

    clarity: float
    atomicity: float
    testability: float
    memorability: float
    accuracy: float

    @property
    def overall(self) -> float:
        """Average across all dimensions."""
        return (
            self.clarity + self.atomicity + self.testability
            + self.memorability + self.accuracy
        ) / 5.0


def assess_quality(*, front: str, back: str) -> QualityScore:
    """Score a card using heuristic checks.

    Each dimension is scored 0.0-1.0:
    - clarity: no vague words, has question structure, not yes/no
    - atomicity: short question, no multi-concept splitting
    - testability: answer is concrete and verifiable
    - memorability: reasonable length, no huge enumerations
    - accuracy: well-formed structure (question mark, non-empty)
    """
    front_lower = front.strip().lower()
    back_stripped = back.strip()

    clarity = _score_clarity(front_lower)
    atomicity = _score_atomicity(front_lower)
    testability = _score_testability(back_stripped)
    memorability = _score_memorability(back_stripped)
    accuracy = _score_accuracy(front.strip(), back_stripped)

    return QualityScore(
        clarity=clarity,
        atomicity=atomicity,
        testability=testability,
        memorability=memorability,
        accuracy=accuracy,
    )


def _score_clarity(front_lower: str) -> float:
    score = 1.0

    for pattern in _VAGUE_PATTERNS:
        if pattern.search(front_lower):
            score -= 0.4
            break

    if any(front_lower.startswith(s) for s in _YES_NO_STARTERS):
        score -= 0.3

    if "?" not in front_lower:
        score -= 0.2

    return max(0.0, score)


def _score_atomicity(front_lower: str) -> float:
    score = 1.0

    word_count = len(front_lower.split())
    if word_count > 30:
        score -= 0.4
    elif word_count > 20:
        score -= 0.2

    matches = _MULTI_CONCEPT_RE.findall(front_lower)
    if len(matches) >= 2:
        score -= 0.4
    elif len(matches) == 1:
        score -= 0.1

    return max(0.0, score)


def _score_testability(back: str) -> float:
    score = 1.0

    word_count = len(back.split())
    if word_count > 200:
        score -= 0.5
    elif word_count > 100:
        score -= 0.3

    if not back:
        return 0.0

    return max(0.0, score)


def _score_memorability(back: str) -> float:
    score = 1.0

    enum_items = len(_ENUM_RE.findall(back))
    if enum_items > 7:
        score -= 0.5
    elif enum_items > 4:
        score -= 0.2

    word_count = len(back.split())
    if word_count > 150:
        score -= 0.3

    return max(0.0, score)


def _score_accuracy(front: str, back: str) -> float:
    score = 1.0

    if not front:
        score -= 0.5
    if not back:
        score -= 0.5

    if front and "?" not in front:
        score -= 0.2

    return max(0.0, score)

File: crates/validation/src/quality.rs
//! Heuristic quality scoring for flashcards.
//!
//! Implements a 5-dimension rubric: clarity, atomicity, testability,
//! memorability, accuracy. Each dimension scores 0.0-1.0.

use serde::Serialize;

/// Five-dimension quality assessment of a flashcard.
#[derive(Debug, Clone, PartialEq, Serialize)]
pub struct QualityScore {
    pub clarity: f64,
    pub atomicity: f64,
    pub testability: f64,
    pub memorability: f64,
    pub accuracy: f64,
}

impl QualityScore {
    /// Average of all five dimensions.
    pub fn overall(&self) -> f64 {
        (self.clarity + self.atomicity + self.testability + self.memorability + self.accuracy) / 5.0
    }
}

/// Score a card using heuristic checks. Each dimension is 0.0-1.0.
///
/// Dimensions:
/// - clarity: penalizes vague openers, yes/no questions, missing "?"
/// - atomicity: penalizes long questions (>20 words), multi-concept splits
/// - testability: penalizes extremely long answers (>100 words), empty answers
/// - memorability: penalizes long enumerations (>4 bullet items), long answers (>150 words)
/// - accuracy: penalizes missing question mark, empty front/back
pub fn assess_quality(front: &str, back: &str) -> QualityScore {
    let front_trimmed = front.trim();
    let back_trimmed = back.trim();

    QualityScore {
        clarity: score_clarity(front_trimmed),
        atomicity: score_atomicity(front_trimmed),
        testability: score_testability(back_trimmed),
        memorability: score_memorability(back_trimmed),
        accuracy: score_accuracy(front_trimmed, back_trimmed),
    }
}

fn score_clarity(front: &str) -> f64 {
    let mut score = 1.0;
    let lower = front.to_lowercase();

    let vague_patterns = [
        "explain",
        "describe",
        "tell me about",
        "tell about",
        "discuss",
        "elaborate",
    ];
    if vague_patterns.iter().any(|p| lower.contains(p)) {
        score -= 0.4;
    }

    let yes_no_starters = [
        "is ", "does ", "can ", "do ", "are ", "was ", "were ", "has ", "have ", "will ", "would ",
        "should ", "could ", "did ",
    ];
    if yes_no_starters.iter().any(|s| lower.starts_with(s)) {
        score -= 0.3;
    }

    if !front.is_empty() && !front.contains('?') {
        score -= 0.2;
    }

    clamp(score)
}

fn score_atomicity(front: &str) -> f64 {
    let mut score = 1.0;
    let word_count = front.split_whitespace().count();

    if word_count > 30 {
        score -= 0.4;
    } else if word_count > 20 {
        score -= 0.2;
    }

    let lower = front.to_lowercase();
    let and_or_count = count_word_boundary(&lower, "and") + count_word_boundary(&lower, "or");

    if and_or_count >= 2 {
        score -= 0.4;
    } else if and_or_count == 1 {
        score -= 0.1;
    }

    clamp(score)
}

fn count_word_boundary(text: &str, word: &str) -> usize {
    let mut count = 0;
    let chars: Vec<char> = text.chars().collect();
    let word_chars: Vec<char> = word.chars().collect();
    let wlen = word_chars.len();

    if chars.len() < wlen {
        return 0;
    }

    for i in 0..=chars.len() - wlen {
        let before_ok = i == 0 || !chars[i - 1].is_alphanumeric();
        let after_ok = i + wlen == chars.len() || !chars[i + wlen].is_alphanumeric();
        let matches = chars[i..i + wlen] == word_chars[..];

        if before_ok && after_ok && matches {
            count += 1;
        }
    }
    count
}

fn score_testability(back: &str) -> f64 {
    if back.is_empty() {
        return 0.0;
    }

    let mut score = 1.0;
    let word_count = back.split_whitespace().count();

    if word_count > 200 {
        score -= 0.5;
    } else if word_count > 100 {
        score -= 0.3;
    }

    clamp(score)
}

fn score_memorability(back: &str) -> f64 {
    let mut score = 1.0;

    let bullet_count = back
        .lines()
        .filter(|line| {
            let trimmed = line.trim_start();
            trimmed.starts_with("- ") || trimmed.starts_with("* ") || is_numbered_item(trimmed)
        })
        .count();

    if bullet_count > 7 {
        score -= 0.5;
    } else if bullet_count > 4 {
        score -= 0.2;
    }

    let word_count = back.split_whitespace().count();
    if word_count > 150 {
        score -= 0.3;
    }

    clamp(score)
}

fn is_numbered_item(line: &str) -> bool {
    let mut chars = line.chars().peekable();

    // Must start with a digit
    if !chars.peek().is_some_and(|c| c.is_ascii_digit()) {
        return false;
    }

    // Consume digits
    while chars.peek().is_some_and(|c| c.is_ascii_digit()) {
        chars.next();
    }

    // Must be followed by ". " or ") "
    match chars.next() {
        Some('.') | Some(')') => chars.next() == Some(' '),
        _ => false,
    }
}

fn score_accuracy(front: &str, back: &str) -> f64 {
    let mut score = 1.0;

    if front.is_empty() {
        score -= 0.5;
    }
    if back.is_empty() {
        score -= 0.5;
    }

    if !front.is_empty() && !front.contains('?') {
        score -= 0.2;
    }

    clamp(score)
}

fn clamp(v: f64) -> f64 {
    v.clamp(0.0, 1.0)
}
